Initial Get-NewPackageFile and Get-RedirectedUri

gjslavik opened the pull request:
Resolves #2. Initial creation of scripts.

Made minor tweaks to the Get-RedirectedUri script that Jeff wrote a while back. Just the order of comparing $null and some of the header info to match functionality correctly. The Get-NewPackageFile does not have any standard output when complete. Is this needed or desired? If so, what type of output should be generated?

Tested Get-RedirectedUri with multiple URIs, they all return as expected.
Tested Get-NewPackageFile with Source and Destination as file, UNC, and URI. Functionality worked as intended, or errors thrown as intended. Function copied files between local directories, from local to UNC, from UNC to UNC, from UNC to local, from URI to local, from URI to UNC, etc.

thoen008 commented on UMN-AutoPackager/public/Get-NewPackageFile.ps1:
```diff
@@ -0,0 +1,141 @@
+the Source parameter is used during the copy. Requires use of the "Get-RedirectedUri" function.
+#>
+    
+    [CmdletBinding(SupportsShouldProcess)]
```
Any reason to not just use [CmdletBinding]()] which would include all the other parameters not just whatif?

thoen008 replied:
Reading up on it it looks like it doesn't enable those without adding that. Resolving the conversation.

jbolduan commented on UMN-AutoPackager/public/Get-RedirectedUri.ps1:
```diff
@@ -0,0 +1,48 @@
+function Get-RedirectedUri {
```
This function needs to be moved to the /private directory.

gjslavik replied:
Will be moved on next update. Resolving conversation.

jbolduan commented on UMN-AutoPackager/public/Get-NewPackageFile.ps1:
```diff
@@ -0,0 +1,141 @@
+
+    process {
+        try {
+            Write-Verbose "Verbose output enabled"
```
You should use the -Message "blah" on Write-Verbose calls to make things clear.  I'm only going to add this comment here, but applies to all instances of Write-Verbose.

I would also clean up the number of Write-Verbose instances you have.  There are way too many.  These execute every time the function is run but are outputted to null if the -Verbose switch isn't enabled.  This still takes up processing time.

gjslavik replied:
Good info! Reducing and consolidating the number of Write-Verbose instances, and adding "-Message" to those that remain. Resolving coversation.

thoen008 reviewed (approved):
Looks good to me.

jbolduan reviewed (changes requested):
I see a couple changes that could/should be made.  Added comments.

gjslavik commented:
@jbolduan I moved Get-RedirectedUri to private per your request. Should Get-NewPackageFile move as well?

gjslavik commented:
Added script functionality to properly handle "WhatIf" and "Confirm" along with "Verbose". Also added a return of "Get-Info" on the saved file so other scripts can work with output if needed.

jbolduan commented on UMN-AutoPackager/private/Get-RedirectedUri.ps1:
```diff
@@ -0,0 +1,48 @@
+function Get-RedirectedUri {
+    <#
+	.SYNOPSIS
+		Gets the real download URL from the redirection.
+	.DESCRIPTION
+		Used to get the real URL for downloading a file, this will not work if downloading the file directly.
+	.EXAMPLE
+		Get-RedirectedURI -URI "https://download.mozilla.org/?product=firefox-latest&os=win&lang=en-US"
+	.PARAMETER URL
+		URL for the redirected URL to be un-obfuscated
+	.NOTES
+		Code from: Redone per issue #2896 in core https://github.com/PowerShell/PowerShell/issues/2896
+	#>
```
We should standardize on help for functions being outside right above the function.

jbolduan commented on UMN-AutoPackager/public/Get-NewPackageFile.ps1:
```diff
@@ -0,0 +1,139 @@
+function Get-NewPackageFile {
+<#
+    .SYNOPSIS
+    Cmdlet that saves a file from a file path, UNC, or URI and saves it to the specified location.
+    .DESCRIPTION
+    Get-NewPackageFile is a function that when provided with a Source file path, UNC, or URI (as a string) and a Destination path or
+    UNC (as a string) will copy or download the Source file to the Destination. Characters of a space (" ") or an escaped space ("%20")
+    are replaced with an underscore ("_") when saved to the Destination. Returns object information of the saved file.
+    .PARAMETER Source
+    A string that must be a file path, UNC, or URI of the file to be saved to the Destination location. Wildcards are not permitted.
+    .PARAMETER Destination
+    A string that must be a file path or UNC where the Source file will be saved. Wildcards are not permitted.
+    .EXAMPLE
+    Get-NewPackageFile -Source "C:\Test\File1.txt" -Destination "D:\Temp\File2.txt"
+    .EXAMPLE
+    Get-NewPackageFile -Source "C:\Test\File1.txt" -Destination "D:\Temp\"
+    .EXAMPLE
+    Get-NewPackageFile -Source "C:\Test\File1.txt" -Destination "\\fileserver\Temp\File1.txt"
+    .EXAMPLE
+    Get-NewPackageFile -Source "\\fileserver\Test\File1.txt" -Destination "D:\Temp\File1.txt"
+    .EXAMPLE
+    Get-NewPackageFile -Source "\\fileserver1\Test\File1.txt" -Destination "\\fileserver2\Temp\File1.txt"
+    .EXAMPLE
+    Get-NewPackageFile -Source "https://webserver.xyz/Test/File1.txt" -Destination "C:\Temp\File1.txt"
+    .EXAMPLE
+    Get-NewPackageFile -Source "https://webserver.xyz/Test/File1.txt" -Destination "\\fileserver\Temp\File1.txt"
+    .INPUTS
+    String
+    .OUTPUTS
+    System.IO.FileInfo
+    Get-NewPackageFile returns information on the file saved in the location specified in the Destination parameter.
+    .NOTES
+    If the Destination parameter is only a file path and does not include a filename, the filename obtained from the Source parameter
+    is used during the copy. Requires use of the "Get-RedirectedUri" function.
+#>
```
See help comment on other function, we should standardize on this being above the function not inside.

jbolduan reviewed (changes requested):
Only changes I think are worth making is moving the help outside the functions.  Otherwise it looks good.

thoen008 reviewed (approved):
I think it was decided to leave the comments under the function name. Otherwise I think everything looks good.

jbolduan reviewed (approved):
We agreed that comment help should be inside the functions based on community coding standards.
